cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SystemProgramManagementService.java

```diff
@@ -0,0 +1,176 @@
+  private Map<ProgramId, Map<String, String>> enabledServices;
+
+  @Inject
+  SystemProgramManagementService(CConfiguration cConf, ProgramRuntimeService programRuntimeService,
+                                 ProgramLifecycleService programLifecycleService) {
+    this.cConf = cConf;
+    this.scheduleInterval = cConf.getLong(Constants.AppFabric.PROGRAM_SCAN_INTERVAL, 10L);
+    this.programRuntimeService = programRuntimeService;
+    this.programLifecycleService = programLifecycleService;
+  }
+
+  /**
+   * Sets the map of services that are currently enabled along with their runtime args.
+   * The services that are not present in map will be stopped during the next run of the service.
+   *
+   * @param enabledServices
+   */
+  public void setEnabledServices(Map<ProgramId, Map<String, String>> enabledServices) {
+    this.enabledServices = new HashMap<>(enabledServices);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    if (enabledServices == null) {
+      LOG.info("Services to run not yet set");
+      return;
+    }
+    try {
+      reconcileServices();
+    } catch (Exception exception) {
+      //catch here so a single iteration doesn't stop the service
+      LOG.error("Error in reconciling services", exception);
+    }
+  }
+
+  private void reconcileServices() {
+    //take a copy
+    Map<ProgramId, Map<String, String>> enabledServicesMap = new HashMap<>(this.enabledServices);
+    Set<ProgramId> enabledServices = enabledServicesMap.keySet();
+    List<ProgramRuntimeService.RuntimeInfo> runtimeInfos = programRuntimeService.listAll(ProgramType.SERVICE);
+    Set<ProgramId> servicesToStop = new HashSet<>();
+    Set<ProgramId> servicesRunning = new HashSet<>();
+    for (ProgramRuntimeService.RuntimeInfo runtimeInfo : runtimeInfos) {
+      ProgramId programId = runtimeInfo.getProgramId();
+      if (!programId.getNamespaceId().equals(NamespaceId.SYSTEM)) {
+        //We care only about system services
+        continue;
+      }
+      if (!enabledServices.contains(programId)) {
+        servicesToStop.add(programId);
+      } else {
+        servicesRunning.add(programId);
+      }
+    }
+    Set<ProgramId> servicesToStart = enabledServices.stream()
+      .filter(programId -> !servicesRunning.contains(programId))
+      .collect(Collectors.toSet());
+    startServices(servicesToStart, enabledServicesMap);
+    stopServices(servicesToStop);
+    //prune additional runs if any
+    pruneAdditionalRuns(enabledServices);
+  }
+
+  private void startServices(Set<ProgramId> servicesToStart, Map<ProgramId, Map<String, String>> enabledServicesMap) {
+    for (ProgramId programId : servicesToStart) {
+      Map<String, String> overrides = enabledServicesMap.get(programId);
+      LOG.debug("Starting program {} with args {}", programId, overrides);
+      try {
+        programLifecycleService.start(programId, overrides, false);
+      } catch (Exception exception) {
```

Comment:
Usually we just use `ex` as the variable name for exception.

Reply:
Done